lib/plugins/aws/lib/naming.js

```diff
@@ -0,0 +1,231 @@
+'use strict';
+
+const _ = require('lodash');
+
+/**
+ * A centralized naming class that provides naming standards for the AWS provider plugin.  The
+ * intent is for this to enable naming standards to be configured using a custom naming plugin or
+ * a naming configuration feature, based on which of those is determined to in the best approach.
+ *
+ * There are two main concepts:
+ * 1. normalization - normalization of a name comprises removing any unacceptable characters and
+ * changing the name to comply with capitalization expectations.
+ * 2. logicalization - logicalization of a name comprises preparing it for use as an attribute name
+ * for a resource within the Service's CloudFormation Template.  Notably, a suffix or prefix should
+ * be added to the name of resources.  These help to disambiguate attributes identifying or
+ * declaring different aspects or data about the same resource.  (e.g. MyFuncLambdaFunction vs.
+ * MyFuncLambdaFunctionArn)  Generally, normalization of the name also occurs.
+ *
+ * Sometimes, it is important to deconstruct the names that are created via this centralized
+ * utility.  As a result, the regular expressions used for identifying specific resource types using
+ * their names or the logic to extract a name components from a logical name are things that must
+ * also be centralized so that they co-occur and are more correlated spatially within the code base.
+ * That is to say that they are easier to change together.
+ */
+class Naming {
+  constructor(serverless) {
+    this.serverless = serverless;
+  }
+
+  /*
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
+  }
+  normalizeNameToAlphaNumericOnly(name) {
+    return this.normalizeName(name.replace(/[^0-9A-Za-z]/g, ''));
+  }
+  normalizePathtoCapitalAlphaNumbericOnlyWithReplacement(path) {
+    return this.normalizeNameToAlphaNumericOnly(
+      path.replace(/-/g, 'Dash')
+          .replace(/\{(.*)\}/g, '$1Var'));
+  }
+
+  // Probably belongs in a non-aws-specific code asset
+  getServiceEndpointRegex() {
+    return /^ServiceEndpoint/;
+  }
+
+  /*
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
```

Comment:
re: "maybe in that case we could just pass the stage/region to whatever method needs them?"
Your proposal is what I was attempting to suggest is problematic.  I could do what you suggest and it **would** fix the **immediate** issue.  It creates a long term problem however.  Doing this constrains what can be used as source material for generating a name (admittedly not an immediate concern).  What about the useful `-foo` switch that will be added to the `deploy` command?  Do we add `foo` to this method signature too?  More plausibly, what about the next provider (beyond AWS) that doesn't have a region and a stage parameter?  Do we add that provider's special CLI configurable values to the signature too?  Over the long term, this approach seems to paint us into a corner.

If I'm missing a big point, I apologize, I'm not sure I quite understand what you mean by 'make those methods "pure" in a sense'.  It sounds as if you're suggesting that the interface be made explicit and thereby that it would be very clear how configuration flows and that would provide a kind of purity.  That said, I don't think that the scope of what will be passed is sufficiently compact or well enough known at this point.

What I am proposing is that centrally SLS will have a scheme for managing the configuration.  It already does this with the `Config` object.  That object could be expanded to provide a richer mechanism of interaction.  The interface of the config object can be made explicit and encapsulated so that data protection occurs and each provider cannot mess with shared state (as others see it) except through well defined interfaces.  Each provider will have effectively unpredictable needs.  As a result, we can centrally maintain the configuration with every provider having an altered version of the centralized configuration available only to them.  I'll get back to this and be concrete.

> as for multiple providers

I am not suggesting that we should provide an abstraction that makes interactions with every provider the same (although similarity would be a little helpful from a cognitive load standpoint).  That leads us toward the least common denominator and none of the deeper value each will have over the others.  Instead, I would suggest we can create a means of separating their perspectives on what the configuration is in isolated contexts that would interact (if ever) through well defined data flows.  Something like:

```
class Config { // the central hive
  constructor(id); // this ID would give you your own version of the configuration
  get(key); // this transparently combines the SLS specified configuration state with the Provider specified configuration state
  set(key, value); // this transparently isolates any value changes by a Provider to the provider
}
```

If it became necessary, you could add more:
```
class Config {
  ...
  getFor(id, key);
  setFor(id, key, value);
}
```
And these would alter the value associated with the given key

```
serverless = new Serverless(); // calls serverless.config.set('foo', 'bar'); and serverless.config.set('my', 'val');
aws = new SDK(serverless);
aws.setFoo('caz') // calls serverless.config.set('foo', 'caz'); // serverless.config now contains global = { foo: 'bar' } and aws = { foo: 'caz' }
serverless.printFoo(); // console.log(this.config.get('foo'); // prints "bar"
aws.printFoo(); // console.log(serverless.config.get('foo'); // prints "caz"
serverless.printMy(); // console.log(this.config.get('my'); // prints "val"
aws.printMy(); // console.log(serverless.config.get('my'); // also prints "val"
```